src/epoch/local.rs

```diff
@@ -0,0 +1,56 @@
+//! Manage the thread-local state.
+//!
+//! This provides access to a `Participant` record.
+
+use std::sync::atomic;
+
+use epoch::participant::Participant;
+use epoch::global;
+
+/// A local epoch state.
+///
+/// The point of this type is mainly the destructor, which should be called on thread exit.
+#[derive(Debug)]
+struct LocalEpoch {
+    /// The particpant of this thread.
+    ///
+    /// This pointer should never be null, but there is currently no way of expressing thread-local
+    /// references ([candidate](https://github.com/rust-lang/rfcs/pull/1705)).
+    participant: *const Participant,
+}
+
+// FIXME: avoid leaking when all threads have exited
+impl Drop for LocalEpoch {
+    fn drop(&mut self) {
+        // Get the participant.
+        let p = unsafe { &*self.participant };
+
+        // Enter the active state.
+        p.enter();
+        // Propagate the garbage to the global state.
+        p.migrate_garbage();
+
+        // Exit the active state.
```

Comment:
The code already _says_ exit the active state. You don't need to repeat it twice.